### ISSUE TITLE
Feature Request: Stable Hints (for scripting)

### ISSUE THREAD

**Issue by C-Loftus** [david-tejada/rango#190]
## Desired Feature 
Allow a specific hint or group of hints on a specific page to be "locked". This would ensure that the hints have the same letters each time, making them predictable and allowing for usage in scripts. Alternatively, we could also have a setting to lock all the hints for the entire page, although this may be more challenging if HTML elements themselves are not consistent across page loads (thus making the lock never come into scope or apply to the new loaded page)

It may also be desirable to change the color of the locked element to distinguish it as such.

## Justification
Currently, there is no efficient way to script browser behavior with Talon without resorting to writing custom JavaScript, creating a browser extension, or manually specifying each Rango target. These methods are time-consuming and not practical for writing scripts, and they do not integrate well with Talon. In contrast, locking specific hints or elements would provide a more streamlined and efficient solution.

This is a better way to script behavior from Talon without needing to pass arbitrary JavaScript.  

## Implementation
With a command like `rango lock <user.rango_target>`, it could be possible to change the class name of the rango-hint `div` to indicate that it is locked. On page load, these locked hints would then be loaded first, and the rest of the page filled in with the rest of the rango_targets. 

It might be helpful for the user to have different locking scopes such as `lock by id <rango_target>` for more strict locking or `lock by class <rango_target>` to get groups of elements at once 

## Other Options
If implementing this feature proves to be difficult, an alternative option could be to provide a command that allows users to copy the underlying HTML, including class and ID information. This would make scripting with JavaScript and Talon substantially easier, eliminating the need to inspect elements and search through raw source code.

## Downsides I can recognize
 The names of rango hints are not semantically useful in talon scripts (since they are just combinations of letters and numbers) and as a result,  the generated scripts might be difficult to share with others who do not share the same locked hint cache. (Perhaps there would be a way to share this down the road? Or alias names for certain hats? Obviously not a priority right now)

**Comment by david-tejada** [comment]
Locking hints to specific elements is more complex than it might appear and not something I'll be able to tackle anytime soon. I will leave this open in case I want to give it a try in the future.

As for the alternative for scripting what you really want is a unique selector. This is much more doable. I'll see if I can implement it when I have some time. I can't promise that it will be soon as I don't have much time to dedicate to this now.

**Comment by david-tejada** [comment]
Answering about [this comment](https://github.com/david-tejada/rango/issues/193#issuecomment-1723609932) in issue #193

> I was thinking of implementing similar behavior (storing unique hints), but for the purpose of scripting. ( For instance, a user could say a command like `mark air bat as input` ) and then use a command to scroll/manipulate `input` where `input` refers to the syntactic name of a hint. Described here: #190
> 
> @SandeepJoel , should we try and coordinate on this? I think there will be some shared functionality related to saving a hint. perhaps it would be good to come to a consensus regarding the best way of saving a hint, and merge that first? I wanted to make sure that we don't have two ways of saving hints, one for scrolling and one for scripting.
> 
> @david-tejada wasn't sure if you have any opinions on how do you want marks to be saved such that they are also exposed to users.
> 
> I am not sure if this is an anti-pattern based on how rango is designed, but I was almost thinking of storing marks (at least for scripting) locally in a JSON file ( assuming it is okay for rango to send an element ID back to Talon over the clipboard and it isn't just one way). That way it is easier to copy and view the mark names when a user is writing scripts locally
> 
> I was thinking of having the command be something like `mark <hint> as <word>` . So saying `save air bat as input` (where `air bat` refers to the input box on slack) could return an updated json with values like the following. Then the user could write a talon script using "input" as one of the input parameters.
> 
> ```
> {
>     "websites": [
>         {
>             "slack": {
>                 "input": "DIV_ID_NAME_GOTTEN_FROM_RANGO"
>             }
>         }
>     ]
> }
> ```
> 
> With regards to scrolling, we could probably just create a dynamic scope in a Talon file, created from that json, instead of needing to save it internally to the extension and obscure it from the user. ( I think we are going to need to have away to export it anyways given the fact that users are likely to switch setups and want to keep their saved marks)
> 
> I suppose I'm just thinking that if we don't save it locally then it is going to be difficult for people to track lots of marks unless we create another page for this which seems like unnecessary work currently.


There is no need to pass json around for this. You just need to store a Map with the key being urls (possibly url patterns like I do for custom selectors) and each value being another Map that relates names to unique selectors. We could simplify this a bit more if we just store the unique selectors in an array. This way each element / unique selector would be represented by its array index. This way we wouldn't need to reference the name/number of the mark when creating them.

The workflow using automatic numbers could be something like this:

`mark element air bat`: Creates and stores the mark. Replaces the hint with the corresponding mark number for a few seconds. Make the hint flash to make it clear.

You would use now this number in talon with something like:

```talon
user.rango_command_without_target("clickElementByMark", 1)
```

Another command could reveal the matching custom markers in a given page. Something like `element marks show`. Here you would just go through all unique selectors and see if the hints in the current page matches any of them and flash the hint with the corresponding index.

If you prefer the named marks you can implement something similar to this but using tooltips (like the ones that show up when you copy a link) instead of changing the text content of the hint. This tooltip approach could also work with the numbered marks solution.

> I think we are going to need to have away to export it anyways given the fact that users are likely to switch setups and want to keep their saved marks

We use sync storage so as long as the user uses the same browser the marks will be kept in sync.

**Comment by C-Loftus** [comment]
That all makes sense. I think I am going to try and implement this with the tooltips and the named marks. My only concern with raw indices is that if there are multiple targets, I feel like the numbered constants start to become difficult (or require extra named constants) to interpret in talon scripts.  Also if we remove a mark, the index will change and potentially mess up the Talon script that doesn't know about that update.
*  realistically some of this stuff probably doesn't matter all that much because I can't imagine these sorts of scripts will be that long,  but I thought it would be better to have named marks generally

So to summarize, just to be on the same page, I think I will try and take a look at this and implement it in a way such that we can do:

`mark <hint> as <word>`. I will then store it in a map with the word as the key and the unique hint id as the value ( It will also have some metadata regarding the url).  Then  we can do  `user.rango_command_without_target("clickElementByMark", $WORD)` .

**Comment by david-tejada** [comment]
I think what it makes most sense is to have something like this, with the objects in curly brackets being maps:

```typescript
{
  "https?://app.slack.com/*": {
    "input": "html > body > div:nth-child(2) > div > div:nth-child(2) > div:nth-child(5) > div:nth-child(2) > div > div:nth-child(4) > div > div:nth-child(2) > div > div > div > div:nth-child(2) > div > div > div > div:nth-child(1) > div",
    "search": "html > body > div:nth-child(2) > div > div:nth-child(2) > div:nth-child(1) > div:nth-child(2) > div:nth-child(3) > button:nth-child(1)"
  },
  "https?://discord.com/*": {
    ...
}
```

To generate the unique selector you can check out [this thread](https://stackoverflow.com/questions/8588301/how-to-generate-unique-css-selector-for-dom-element). [This library](https://www.npmjs.com/package/css-selector-generator) looks good too.

**Comment by david-tejada** [comment]
Forgot to mention there's already a function for getting the host pattern. Search for `getHostPattern`. You can extract it to a more generic place if you use it
